feat(appset): implement ResourceTree for ApplicationSets

alexymantha opened the pull request:
<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Part of the effort to add support for ApplicationSets in the UI, adds a ResourceTree endpoint similar that will be used to display the tree of an ApplicationSet and its children Applications. 

As discussed in the linked issue, this uses the ApplicationSet status to track the children and the tree is built based on the data contained in the status.

The progressive sync feature added an ApplicationSetApplicationStatus field that is used to track the status and progress of the sync of the children's applications. I tried to reuse what the progressive sync was already doing. See comments for more details on the refactor.

Closes #15697 

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them. (UI will be implemented in another PR)
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

reggie-k commented:
Thanks for being so quick with this!
Can you elaborate about the progressive sync changes?

alexymantha commented:
> Can you elaborate about the progressive sync changes?

Yes! I tried to reuse the ApplicationSetApplicationStatus that was previously only used for the progressive sync so I had to make some changes to the progressive sync:
- moved some of the logic that was handling the status update outside of the feature flag.
- moved the progressivesync-specific fields to a subfield of the ApplicationSetApplicationStatus.
- change to use a status map instead of a slice to make it easier to update the status along the way since it's not only being updated by the progressive sync anymore.

Most of these changes could be avoided if I didn't try to reuse ApplicationSetApplicationStatus but I think it makes sense to make the status more generic and not depend on the progressive sync.

reggie-k commented:
Thanks.
Let's wait for Michael's review.
And I agree that only nodes are relevant for AppSet, orphan nodes are not.

On Sun, Oct 1, 2023, 08:19 Alexy Mantha ***@***.***> wrote:

> Can you elaborate about the progressive sync changes?
>
> Yes! I tried to reuse the ApplicationSetApplicationStatus that was
> previously only used for the progressive sync so I had to make some changes
> to the progressive sync:
>
>    - moved some of the logic that was handling the status update outside
>    of the feature flag.
>    - moved the progressivesync-specific fields to a subfield of the
>    ApplicationSetApplicationStatus.
>    - change to use a status map instead of a slice to make it easier to
>    update the status along the way since it's not only being updated by the
>    progressive sync anymore.
>
> Most of these changes could be avoided if I didn't try to reuse
> ApplicationSetApplicationStatus but I think it makes sense to make the
> status more generic and not depend on the progressive sync.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/argoproj/argo-cd/pull/15741#issuecomment-1741962637>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEVDWBGTB6OFNUX4NDEHJEDX5D4MVANCNFSM6AAAAAA5NG7IHE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

reggie-k commented:
@alexymantha is this waiting for a review or still in progress?

crenshaw-dev commented on pkg/apis/application/v1alpha1/applicationset_types.go:
```diff
@@ -783,7 +784,15 @@ const (
+
+	ProgressiveSync *ProgressiveSyncStatus `json:"progressiveSyncStatus,omitempty" protobuf:"bytes,4,opt,name=progressiveSyncStatus"`
+	Health          HealthStatus           `json:"health,omitempty" protobuf:"bytes,5,opt,name=health"`
+}
+
+type ProgressiveSyncStatus struct {
```
Would it be possible to flatten this back out? Nesting the progressive sync fields introduces a world of backwards-compatibility pain.

alexymantha replied:
Sure! However, is there another way to clarify what fields are specifically for the progressive sync? I fear the `Status` and `Message` that are specifically for the progressive sync will get confused with the App's `Status` and `Message` which are under `Health`

crenshaw-dev replied:
You could prefix the field names with `ProgressiveSync` but leave the `json` tags as-is. That would clarify the Go code without breaking the k8s or Argo API contracts.

alexymantha replied:
Done!

crenshaw-dev commented on pkg/apis/application/v1alpha1/applicationset_types.go:
```diff
@@ -783,15 +784,21 @@ const (
+
+  // Progressive Sync fields
+	// ProgressiveSyncLastTransitionTime is the time the status was last updated
+	ProgressiveSyncLastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,5,opt,name=lastTransitionTime"`
+	// ProgressiveSyncMessage contains human-readable message indicating details about the status of the progressive sync
+	ProgressiveSyncMessage string `json:"message" protobuf:"bytes,6,opt,name=message"`
+	// ProgressiveSyncStatus contains the AppSet's perceived status of the managed Application resource: (Waiting, Pending, Progressing, Healthy)
+  // this is NOT the same as the Application's status
+	ProgressiveSyncStatus string `json:"status" protobuf:"bytes,7,opt,name=status"`
+	// ProgressiveSyncStep tracks which step the Applications should be updated in
+	ProgressiveSyncStep string `json:"step" protobuf:"bytes,8,opt,name=step"`
```
To avoid a breaking gRPC change, keep the number in the `protobuf` tag for each of these the same and add the new fields after these. It might read a bit wonky, but worth it to avoid the breaking API change.

alexymantha replied:
Ah ok thanks for the info! I was wondering if it had an impact

alexymantha replied:
Done!

alexymantha commented on applicationset/controllers/applicationset_controller_test.go:
```diff
@@ -4736,69 +4736,12 @@ func TestUpdateApplicationSetApplicationStatus(t *testing.T) {
-			},
-		},
-		{
-			name: "removes the appStatus for applications that no longer exist",
```
Responsibility of removing the status moved out of the progressive sync

alexymantha commented on applicationset/controllers/applicationset_controller_test.go:
```diff
@@ -5889,3 +5838,267 @@ spec:
 	}
 }
+
+func TestAddStatus(t *testing.T) {
```
New tests start here, the rest is just the renaming of the progressive sync fields and change to status map

alexymantha commented:
I think the PR is ready for feedback. I'm also working on adding an E2E to test the status updates

reggie-k commented:
@alexymantha can you update this branch from upstream master?

alexymantha commented:
@reggie-k It should be good to go!

reggie-k commented:
Thanks a lot! I will test the changes.
As per the discussion with @crenshaw-dev, I will merge these changes into
my original PR, also the other PRs you are working on, so that everything
gets reviewed at once.

Can you also add changes in the CLI for the AppSet resourc-tree?

On Thu, Jan 4, 2024, 19:05 Alexy Mantha ***@***.***> wrote:

> Hi @reggie-k <https://github.com/reggie-k>! I added the missing group and
> kind, can you confirm if all the missing info is there? If so, this PR
> should be ready for a review.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/argoproj/argo-cd/pull/15741#issuecomment-1877460879>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEVDWBARUU4ZJTU7STBMZKTYM3OMLAVCNFSM6AAAAAA5NG7IHGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNZXGQ3DAOBXHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

reggie-k commented:
As per discussion with @crenshaw-dev, relevant CLI changes can be added later on separately.

reggie-k commented:
Verified - the missing fields are now present, working as expected.
Thanks @alexymantha!

csantanapr commented:
what are the next steps to merge this @alexymantha and @reggie-k, just curious on this because there is another PR I'm interested https://github.com/argoproj/argo-cd/pull/16781

alexymantha commented:
IIUC @crenshaw-dev suggested to merge it alongisde https://github.com/argoproj/argo-cd/pull/15731 to make it easier for the reviews but this PR is standalone and could be merged now if needed. It's up to the maintainers

todaywasawesome reviewed (approved):
@crenshaw-dev I recommend moving ahead with this PR standalone unless you feel differently. 

Sorry @alexymantha it needs merge conflicts fixed again.

alexymantha commented:
> Sorry @alexymantha it needs merge conflicts fixed again.

Done!

Tyrael commented:
hi, what is the status here? any reason why this isn't merged yet?

crenshaw-dev reviewed (approved):
Code looks good, local test checks out. Thanks, @alexymantha!

crenshaw-dev commented:
Would you mind rebasing and re-running codegen?

alexymantha commented:
Thanks @crenshaw-dev! Everything is up-to-date now

Tyrael commented:
Thanks!

reggie-k commented:
Thank you @alexymantha and @crenshaw-dev!
